src/StepControl.h

```diff
@@ -1,72 +1,78 @@
+#include <cmath>
 #include "StepControlBase.h"
 
-class StepControl : public StepControlBase<>
-{
-  protected:
-    uint32_t accelerationEnd;
-    uint32_t decelerationStart;
-    uint32_t sqrt_2a;
-    uint32_t s_tgt, v_tgt, v_min;
+template <unsigned pulseWidth = 5, unsigned accUpdatePeriod = 5000>
+class StepControl : public StepControlBase<pulseWidth, accUpdatePeriod>
+{  
+    int32_t s_0;
+    uint32_t delta_tgt;
+    uint32_t accEnd, decStart;
+    uint32_t two_a;
+    uint32_t v_tgt, v_min2;
 
-    inline uint32_t prepareMovement(uint32_t targetPos, uint32_t targetSpeed, uint32_t pullInSpeed, uint32_t acceleration);
-    inline uint32_t updateSpeed(uint32_t currentPosition);
-    inline uint32_t initiateStopping(uint32_t currentPosition);
+    inline uint32_t prepareMovement(int32_t currentPos, int32_t targetPos, uint32_t targetSpeed, uint32_t acceleration);
+    inline uint32_t updateSpeed(int32_t currentPos);
+    inline uint32_t initiateStopping(int32_t currentPos);
 };
 
-uint32_t StepControl::prepareMovement(uint32_t targetPos, uint32_t targetSpeed, uint32_t pullInSpeed, uint32_t a)
+template <unsigned p, unsigned u>
+uint32_t StepControl<p, u>::prepareMovement(int32_t currentPos, int32_t targetPos, uint32_t targetSpeed, uint32_t a)
 {
-    s_tgt = targetPos;
+    s_0 = currentPos;
+    delta_tgt = std::abs(targetPos - currentPos);
     v_tgt = targetSpeed;
-    v_min = pullInSpeed;
-
-    if (v_tgt > v_min)                   // target speed larger than pull in speed
-    {                                    // -> acceleration required
-        float dv = v_tgt - v_min;
-        float ae = dv * dv / (2.0f * a); // length of acceleration phase (steps)
-        sqrt_2a = sqrtf(2.0f * a);       // precalc, needed during acceleration
+    two_a = v_min2 = 2 * a;                                 // v^2 - v0^2 = 2as -> v0^2 = 2as  (1)
+    uint32_t v_min = sqrtf(v_min2);
```

Comment:
Makes sense. Any value less than 2a will meet the acceleration requirement.

The ability to set the minimum velocity is still useful when using paths. For instance if it's known a constant arc or curve is coming up the minimum and maximum can be used to "clamp" the speed so it moves at a constant rate along the curve instead of stopping at each point in the curve.